backend/database.py

```
@@ -247,6 +247,56 @@ def set_preference(key: str, value: str):
+    Callers should deduplicate input indices before calling. Raises ValueError
+    if more than MAX_INDICES unique indices are provided to prevent SQLite
+    parameter overflow.
+    """
+    if not faiss_indices:
+        return {}
+
+    unique_indices = list(dict.fromkeys(faiss_indices))
+    if len(unique_indices) > MAX_INDICES:
+        raise ValueError(
+            f"Too many indices: {len(unique_indices)} exceeds MAX_INDICES={MAX_INDICES}"
+        )
```

Ruff (0.15.1)

[warning] Avoid specifying long messages outside the exception class

(TRY003)